electrum/onion_message.py

```diff
@@ -0,0 +1,759 @@
+# Electrum - Lightweight Bitcoin Client
+# Copyright (c) 2023-2024 Thomas Voegtlin
+#
+# Permission is hereby granted, free of charge, to any person
+# obtaining a copy of this software and associated documentation files
+# (the "Software"), to deal in the Software without restriction,
+# including without limitation the rights to use, copy, modify, merge,
+# publish, distribute, sublicense, and/or sell copies of the Software,
+# and to permit persons to whom the Software is furnished to do so,
+# subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be
+# included in all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+# BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+# ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+# CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+import asyncio
+import copy
+import io
+import os
+import threading
+import time
+from random import random
+
+from typing import TYPE_CHECKING, Optional, Sequence, NamedTuple
+
+import electrum_ecc as ecc
+
+from electrum.lnrouter import PathEdge
+from electrum.logging import get_logger, Logger
+from electrum.crypto import sha256, get_ecdh
+from electrum.lnmsg import OnionWireSerializer
+from electrum.lnonion import (get_bolt04_onion_key, OnionPacket, process_onion_packet,
+                              OnionHopsDataSingle, decrypt_onionmsg_data_tlv, encrypt_onionmsg_data_tlv,
+                              get_shared_secrets_along_route, new_onion_packet)
+from electrum.lnutil import LnFeatures
+from electrum.util import OldTaskGroup, log_exceptions
+
+
+# do not use util.now, because it rounds to integers
+def now():
+    return time.time()
```

Comment:
Ok, but then `util.now` is not used anywhere? :)
Different use cases have different needs for "now", so maybe it does not make sense to have a `util.now`.

Some use cases might want higher than second precision, others might want `time.monotonic()`, etc.

(In fact you might want to use `time.monotonic()` in this file because looks like only time delays are being measured, absolute time is not used. But anyway. We can always improve this later.)

Reply:
`util.now` was introduced in this PR, but left in after changing to `onion_message.now`. I'll remove it from the PR.